semgrep-core/synthesizing/Pattern_from_Code.ml

```diff
@@ -28,25 +28,146 @@ open AST_generic
 type named_variants =
   (string * Pattern.t) list
 
+type global_state = { count : int; mapping : (expr * expr) list;
+                      has_type : bool }
+
+(*****************************************************************************)
+(* State helpers *)
+(*****************************************************************************)
+
+(* TODO make mapping a map and use map lookup *)
+let lookup state e =
+let mapping = state.mapping in
+let rec look = function
+    | [] -> None
+    | (e1, e2)::xs ->
+      if Matching_generic.equal_ast_binded_code (E e) (E e1) then Some e2 else look xs
+in
+  look mapping
+
+
 (*****************************************************************************)
 (* Helpers *)
 (*****************************************************************************)
 let fk = Parse_info.fake_info "fake"
 let _bk f (lp,x,rp) = (lp, f x, rp)
 
+let default_id str =
+  Id((str, fk),
+   {id_resolved = ref None; id_type = ref None; id_const_literal = ref None})
+
+let default_tyvar str typ =
+  TypedMetavar((str, fk), fk, typ)
+
+let count_to_id count =
+  let make_id ch = Format.sprintf "$%c" ch in
+  if count = 1 then make_id 'X'
```

Comment:
you can write this as
match count with
| 1 -> make_id 'X'
| 2 -> make_id 'Y'
| _ when count <= 26 -> ...
| _ -> Format.sprintf ...